certs/certs.go

```diff
@@ -44,365 +41,159 @@ type LoadX509KeyPairFunc func(certFile, keyFile string) (tls.Certificate, error)
-// When a client tries to establish a TLS connection, Manager will try to
-// pick a certificate that can be validated by the client.
-//
-// For instance, if the client specifies a TLS SNI then Manager will try to
-// find the corresponding certificate. If there is no such certificate it
-// will fallback to the certificate named public.crt.
-//
-// Manager will automatically reload certificates if the corresponding file changes.
-type Manager struct {
-	lock         sync.RWMutex
-	certificates map[pair]*tls.Certificate // Mapping: certificate file name => TLS certificates
-	defaultCert  pair
-	duration     time.Duration
-
+// Certificate is a chain of one or more reloadable certificates.
+type Certificate struct {
+	certFile        string
+	keyFile         string
 	loadX509KeyPair LoadX509KeyPairFunc
-	done            <-chan struct{}
-	reloadCerts     []chan struct{}
-}
 
-var isk8s = env.Get("KUBERNETES_SERVICE_HOST", "") != ""
+	lock        sync.RWMutex
+	certificate tls.Certificate
 
-// pair represents a certificate and private key file tuple.
-type pair struct {
-	KeyFile  string
-	CertFile string
+	listenerLock sync.Mutex
+	listeners    []chan<- tls.Certificate
 }
 
-// NewManager returns a new Manager that handles one certificate specified via
-// the certFile and keyFile. It will use the loadX509KeyPair function to (re)load
-// certificates.
-//
-// The certificate loaded from certFile is considered the default certificate.
-// If a client does not send the TLS SNI extension then Manager will return
-// this certificate.
-func NewManager(ctx context.Context, certFile, keyFile string, loadX509KeyPair LoadX509KeyPairFunc) (manager *Manager, err error) {
-	certFile, err = filepath.Abs(certFile)
+// NewCertificate returns a new Certificate from the given certficate and private key file.
+// On a reload event, the certificate is reloaded using the loadX509KeyPair function.
+func NewCertificate(certFile, keyFile string, loadX509KeyPair LoadX509KeyPairFunc) (*Certificate, error) {
+	certFile, err := filepath.Abs(certFile)
 	if err != nil {
 		return nil, err
 	}
 	keyFile, err = filepath.Abs(keyFile)
 	if err != nil {
 		return nil, err
 	}
-
-	manager = &Manager{
-		certificates: map[pair]*tls.Certificate{},
-		defaultCert: pair{
-			KeyFile:  keyFile,
-			CertFile: certFile,
-		},
+	c := &Certificate{
+		certFile:        certFile,
+		keyFile:         keyFile,
 		loadX509KeyPair: loadX509KeyPair,
-		done:            ctx.Done(),
-		duration:        1 * time.Minute,
 	}
-	if err := manager.AddCertificate(certFile, keyFile); err != nil {
+	if err := c.Reload(); err != nil {
 		return nil, err
 	}
-	return manager, nil
+	return c, nil
 }
 
-// UpdateReloadDuration set custom symlink reload duration
-func (m *Manager) UpdateReloadDuration(t time.Duration) {
-	m.lock.Lock()
-	m.duration = t
-	m.lock.Unlock()
+// Certificate returns the current TLS certificate.
+func (c *Certificate) Certificate() tls.Certificate {
```

Comment:
```suggestion
func (c *Certificate) Get() tls.Certificate {
```

Shouldn't this be just `Get()` ?